libs/nucs/src/selector.rs

```diff
@@ -1,17 +1,58 @@
-use std::{collections::HashSet, fmt, str::FromStr, sync::LazyLock};
+use std::{
+    collections::{HashMap, HashSet},
+    fmt,
+    str::FromStr,
+    sync::LazyLock,
+};
 
 use serde_with::{DeserializeFromStr, SerializeDisplay};
 
 static LABEL_ALPHABET: LazyLock<HashSet<char>> =
     LazyLock::new(|| ('a'..='z').chain('A'..='Z').chain('0'..='9').chain(['_', '-']).collect());
 
+/// The target of a selector.
+#[derive(Clone, Debug, PartialEq)]
+enum SelectorTarget {
+    /// The target is the token itself.
+    Token,
+
+    /// The target is the external context.
+    Context,
+}
+
 /// A selector that can be applied to a NUC.
 #[derive(Clone, Debug, PartialEq, SerializeDisplay, DeserializeFromStr)]
-pub struct Selector(Vec<String>);
+pub struct Selector {
+    labels: Vec<String>,
+    target: SelectorTarget,
+}
 
 impl Selector {
-    pub fn apply<'a>(&self, mut value: &'a serde_json::Value) -> &'a serde_json::Value {
-        for label in &self.0 {
+    /// Apply this selector on a value and context and return the selected value.
+    pub fn apply<'a>(
+        &self,
+        value: &'a serde_json::Value,
+        context: &'a HashMap<&str, serde_json::Value>,
+    ) -> &'a serde_json::Value {
+        match &self.target {
+            SelectorTarget::Token => Self::apply_on_value(&self.labels, value),
+            SelectorTarget::Context => match self.labels.first() {
+                Some(label) => {
+                    // Use the first label look up which entry in the context we should be using.
+                    let Some(value) = context.get(label.as_str()) else {
```

Comment:
What if selector provided by user is incorrect, and looks for an object, but actual value is a scalar?
Ideally we should also check for value type and return error that value is not an object, instead of Null?

Reply:
This situation already existed in the previous code and that's fine. If you have a JSON like `{"foo": 42}` and you have a policy like `["eq", ".foo", "potato"]` the policy will fail because equality is false (`42 != "potato"`). The selector only returns the thing you want or nothing (null) if it can't find it. Similar if for the same JSON you tried to do `["eq", ".foo.bar", 1337]` it will fail as expected. 

Now with the context there's another lookup above that but the behavior should be the same. If your context map is a single key "bar" that has `{"foo": 42}` inside, then this should fail `["eq", "$.bar.missing", 1337]` and `["eq", "$.missing", 1337]` should fail too.

Note that by "fail" I mean "the policy evaluation will return false and the token will be considered invalid.

Reply:
bear in mind I am not aware of full end-to-end error processing here, maybe its to complex to report this kind of error during policy eval, so I am just suggesting this.

Reply:
I see yeah, there is no such thing as a "bad selector" (as long as it's well formed) while you're evaluating a policy. The selector is only used to extract information out of a token to apply it to a policy. The policy evaluation returns true/false, there's nothing outside of that. If you provide a policy that expects something and that's not met, it return false. Think of this as a very restricted scripting language, e.g. as if this was JS and you did

```js
const obj = {
    a: 42
};

// Both are fine and return `false`
obj.a == "boop"; // exists but is different
obj.b == "boop"; // does not exist
```